smartsim/_core/_cli/build.py

```diff
@@ -285,73 +223,236 @@ def build_redis_ai(
+        },
+        install_on_absent=fetch_if_missing,
+        install_on_conflict=False,
+        verbose=verbose,
+    )
+
+
+def install_py_tf_version(
+    fetch_if_missing: bool = False,
+    verbose: bool = False,
+) -> None:
+    """Check Python environment for a compatible TensorFlow installation"""
+    logger.info("Searching for a compatible TF install...")
+    _confirm_package_in_python_env(
+        {"tensorflow": Versioner.TENSORFLOW},
+        install_on_absent=fetch_if_missing,
+        install_on_conflict=False,
+        verbose=verbose,
+    )
+
+
+def _confirm_package_in_python_env(
+    packages: t.Mapping[str, t.Optional[Version_]],
+    package_pinning: _TPinningStr = "==",
+    end_point: t.Optional[str] = None,
+    validate_installed_version: t.Optional[
+        t.Callable[[str, t.Optional[Version_]], bool]
+    ] = None,
+    install_on_absent: bool = False,
+    install_on_conflict: bool = False,
+    verbose: bool = False,
+) -> None:
+    # TODO: Do not like how the defualt validation function will always look for
+    #       a `==` pinning. Maybe turn `BuildEnv.check_installed` into a factory
+    #       that takes a pinning and an appropiate validation fn?
+    validate_installed_version = validate_installed_version or BuildEnv.check_installed
+    missing, conflicts, to_install, to_uninstall = _assess_python_env(
+        packages,
+        package_pinning,
+        validate_installed_version,
+        install_on_absent,
+        install_on_conflict,
+        verbose=verbose,
+    )
+
+    if missing or conflicts:
+        indent = "\n\t"
+        fmt_list: t.Callable = lambda n, l: f"{n}:{indent}{indent.join(l)}" if l else ""
+        missing_str = fmt_list("Missing", missing)
+        conflict_str = fmt_list("Conflicting", conflicts)
+        sep = "\n" if missing_str and conflict_str else ""
+        logger.warning(
+            "Python Env Status Warning!\n"
+            "Requested Packages are Missing or Conflicting:\n\n"
+            f"{missing_str}{sep}{conflict_str}"
+            "\n\nConsider installing packages at the requested versions via "
+            "`pip` or re-running `smart build` with `--modify_python_env`"
+        )
+    if to_uninstall:
+        pip_uninstall(to_uninstall, verbose=verbose)
+    if to_install:
+        pip_install(to_install, end_point=end_point, verbose=verbose)
+
+
+def _assess_python_env(
+    packages: t.Mapping[str, t.Optional[Version_]],
+    package_pinning: _TPinningStr,
+    validate_installed_version: t.Callable[[str, t.Optional[Version_]], bool],
+    install_on_absent: bool,
+    install_on_conflict: bool,
+    verbose: bool,
+) -> t.Tuple[t.List[str], t.List[str], t.List[str], t.List[str]]:
+    missing: t.List[str] = []
+    conflicts: t.List[str] = []
+    to_uninstall: t.List[str] = []
+    to_install: t.List[str] = []
+
+    verbose_info: t.Callable[[str], t.Any] = lambda s: ...
```

Comment:
this looks like something you could pull out into "global space" and reuse. Are we planning on adding any verbose logging elsewhere that we could gain from not repeating these lines?

Reply:
I'm thinking this might also be cause for a custom log level instead of this extra set of declarations?

Reply:
Love the concept of a custom "verbose log" log level, but a little bit worried about squeezing into this ticket. Maybe make a new one?